### DISCUSSION TITLE
UniValue Thrown

### DISCUSSION THREAD

**Discussion by spflanze** [pocketnetteam/pocketnet.core#705] [Q&A]
My new node was synchronizing fine for 3 days. But now, with two years of node data to go, I can no longer start up synchronization. Below is the response I get:

```
~$ pocketcoin-qt
0000ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff
0000ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff


************************
UNKNOWN EXCEPTION       
pocketcoin in loadblk       

terminate called after throwing an instance of 'UniValue'
Aborted (core dumped)
```
How is this fixed?

**Comment by andyoknen** [comment]
Version is 0.22.4 ?

**Reply by spflanze** [reply]
Yes. The installation filename is:
pocketnetcore_0.22.4_linux_x64_setup.deb
This was downloaded on April, 16, 2024.

**Reply by andyoknen** [reply]
Maybe you can run the node in debugging?

```sh
> gdb pocketcoin-qt
...
enter "r" for start
```

**Reply by spflanze** [reply]
The result of the debug run:

```
(gdb) r
Starting program: /usr/local/bin/pocketcoin-qt 
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
[New Thread 0x7ffff74ca640 (LWP 127781)]
0000ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff
0000ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff
[New Thread 0x7ffff64c7640 (LWP 127783)]
[New Thread 0x7ffff5948640 (LWP 127784)]
[New Thread 0x7fffe6ffe640 (LWP 127785)]
[New Thread 0x7fffe67fd640 (LWP 127786)]
[New Thread 0x7fffe5ffc640 (LWP 127787)]
[New Thread 0x7fffe57fb640 (LWP 127788)]
[New Thread 0x7fffe4ffa640 (LWP 127789)]
[New Thread 0x7fffd7fff640 (LWP 127790)]
[New Thread 0x7fffd77fe640 (LWP 127791)]
[New Thread 0x7fffd6ffd640 (LWP 127792)]
[New Thread 0x7fffd67fc640 (LWP 127793)]
[New Thread 0x7fffd5ffb640 (LWP 127794)]
[New Thread 0x7fffd57fa640 (LWP 127795)]
[New Thread 0x7fffd4ff9640 (LWP 127796)]
[New Thread 0x7fffb3fff640 (LWP 127797)]
[New Thread 0x7fffb37fe640 (LWP 127798)]
[New Thread 0x7fffb2ffd640 (LWP 127799)]
[New Thread 0x7fffb27fc640 (LWP 127800)]
[New Thread 0x7fffb1ffb640 (LWP 127804)]
[New Thread 0x7fffb17fa640 (LWP 127805)]
[New Thread 0x7fffb0ff9640 (LWP 127806)]
[New Thread 0x7fff93fff640 (LWP 127807)]
[New Thread 0x7fff937fe640 (LWP 127808)]
[New Thread 0x7fff92ffd640 (LWP 127809)]
[New Thread 0x7fff927fc640 (LWP 127810)]
[New Thread 0x7fff91ffb640 (LWP 127811)]
[New Thread 0x7fff917fa640 (LWP 127812)]
[New Thread 0x7fff90ff9640 (LWP 127813)]
[New Thread 0x7fff6ffff640 (LWP 127814)]
[New Thread 0x7fff6f7fe640 (LWP 127815)]
[New Thread 0x7fff6effd640 (LWP 127816)]
[New Thread 0x7fff6e7fc640 (LWP 127817)]
[New Thread 0x7fff6dffb640 (LWP 127818)]
[New Thread 0x7fff6d7fa640 (LWP 127819)]
[New Thread 0x7fff6cff9640 (LWP 127820)]
[New Thread 0x7fff4ffff640 (LWP 127821)]
[New Thread 0x7fff4f7fe640 (LWP 127822)]
[New Thread 0x7fff4effd640 (LWP 127823)]
[New Thread 0x7fff4e7fc640 (LWP 127824)]
[New Thread 0x7ffec6146640 (LWP 128167)]
[New Thread 0x7ffec5945640 (LWP 128168)]
[New Thread 0x7ffec5144640 (LWP 128169)]
[New Thread 0x7ffec4943640 (LWP 128170)]
[New Thread 0x7ffeb7fff640 (LWP 128171)]
[Thread 0x7ffec4943640 (LWP 128170) exited]
[New Thread 0x7ffeb77fe640 (LWP 128172)]
[New Thread 0x7ffeb6ffd640 (LWP 128173)]
[New Thread 0x7ffeb67fc640 (LWP 128174)]
[New Thread 0x7ffeb5ffb640 (LWP 128175)]
[New Thread 0x7ffeb57fa640 (LWP 128176)]


************************
UNKNOWN EXCEPTION       
pocketcoin in loadblk       

terminate called after throwing an instance of 'UniValue'
[New Thread 0x7ffeb4bd3640 (LWP 128182)]

Thread 42 "b-loadblk" received signal SIGABRT, Aborted.
[Switching to Thread 0x7ffec6146640 (LWP 128167)]
__pthread_kill_implementation (no_tid=0, signo=6, threadid=140732221646400) at ./nptl/pthread_kill.c:44
44	./nptl/pthread_kill.c: No such file or directory.
(gdb) 
```

**Reply by andyoknen** [reply]
This gives a little more information. I would also like to ask you to delete the debug.log file and run it with the argument:
```sh
> pocketcoin-qt -debug=all
```

Most likely, some logging function has encountered nullptr - you can try to track the sequence of actions using detailed logs.

**Reply by spflanze** [reply]
I did  your recommended commands. The last lines of the debug.log file which appear to be relevant are:

```
2024-05-20T14:51:01Z 11; Message: database corruption at line 70719 of [831d0fb283]
2024-05-20T14:51:01Z 11; Message: database corruption at line 70893 of [831d0fb283]
2024-05-20T14:51:01Z 11; Message: statement aborts at 77: [
            update Ratings indexed by Ratings_Type_Uid_Last_Height
                set Last = 0
            where Type = ?
                and Last = 1
                and Uid = ?
   
2024-05-20T14:51:01Z 

************************
UNKNOWN EXCEPTION       
pocketcoin in loadblk
```
I believe the corruption happened when I forgot to shut down synchronization before shutting down my computer. Can this database be fixed? Or will I have to start over? If I were to start over, what files would have to be deleted, or what command would be executed?

I will attempt to prevent this corruption happening again by putting the "pocketcoin-cli stop" command in a script that runs just before shutdown by following instructions in this thread:
https://askubuntu.com/questions/1449198/running-a-script-at-shutdown-before-networking-goes-down
To prevent corruption is it necessary to shutdown pocketcoin before the network goes down? Or is it sufficient to do it before the OS shuts down?
If you have a better way to automate shutting down pocketcoin please suggest it.

**Reply by andyoknen** [reply]
Sorry for the long answer.

To download the blockchain faster, use a snapshot: https://github.com/pocketnetteam/pocketnet.core/blob/0.22/README.md#initialize-blockchain-data-with-database-checkpoint

If you turn on staking, then in the absence of a network (Internet) your node can fork without active peers (it will start making blocks, because the complexity of the network will drop). To avoid this, you can enable the option below - then your node will wait for the connection of the minimum number of peers to continue staking. In other cases, disconnecting the Internet is not critical.

> line in `pocketcoin.conf`
```
stakingrequirespeers=1
```


Regarding stopping the node before shutting down the PC, it seems like a good idea. I can also recommend that you use a docker if possible - in most cases, the docker will take care of a "smooth" stop of the container when possible.

https://github.com/pocketnetteam/pocketnet.core/blob/0.22/README.md#docker

**Reply by spflanze** [reply]
Thank you for your reply.

If I were to to do the initialize blockchain data with database checkpoint procedure given in your link will it overwrite the years of data I have already downloaded during synchronization before the corruption happened? Or should I delete the node data synchronized so far before doing this?

Should "pocketnetcore_0.22.4_linux_x64_setup.deb" installation be removed before installing docker? Or is docker an additional installation?

**Reply by andyoknen** [reply]
When uploading a snapshot, it is necessary to delete directories with the same name from the working directory of the node.

The snapshot contains the structure and data for all years. The database has been cleared of indexes and cache. The first launch will take up to a day.

Docker is a container management system, aka a virtual machine. In practice, running multiple nodes does not cause failures in all instances. There may be difficulties with maintaining network ports and additional disk space

**Reply by spflanze** [reply]
I do not have a file named "pocketcoin.conf" on my computer. I do have this file name and path:
`/home/stephen/.config/Pocketcoin/Pocketcoin-Qt.conf`
Is that it?

Where will I find the files that need to be deleted? Is there a pocketcoin command that will tell me this?

**Reply by andyoknen** [reply]
The default path for linux is `/home/USERNAME/.pocketcoin`

**Reply by spflanze** [reply]
Thank you andyoknen for you rely. You have been a big help so far.

In my .pocketcoin folder are the directories:
`blocks  chainstate  checkpoints  database  pocketdb  wwwroot`

The directories: "checkpoints", and "wwwroot", are empty. 

In the "blocks" directory are many numbered blk*.dat files, and in it's "index" subdirectory, many numbered *.ldb files.

In the "chainstate" directory are many numbered *.ldb files.

In the "database" directory the only file is: "log.0000000001"

In the "pocketdb" directory are many files with "sqlite3" in their names

In the ".pocketcoin/" root are the files:
`banlist.dat db.log debug.log fee_estimates.dat mempool.dat peers.dat pocketcoind.pid settings.json wallet.dat`

Which of the above files should be deleted? I can surmise for example *.ldb files would be deleted, and anything in the "pocketdb" directory not. But I am concerned I will eventually guess wrong and corrupt something.

**Reply by andyoknen** [reply]
Important files that cannot be recovered are:
 - file `wallet.dat` in root directory
 - directory `wallets` if exists
 - file `pocketcoin.conf` contains your configuration options

All other files and directories have no value.

**Reply by spflanze** [reply]
Thank you for your reply. I should have read the snapshot download documents sooner. In them are the directories that should be deleted. I have it downloaded.

Included in the download is the file "pocketcoin.conf.example". I copied this to the filename "pocketcoin.conf", and added your recommended "stakingrequirespeers=1" line to it.

I now have a pocketcoin node running on my computer.

My concern now is automated node shutdown when my computer is shutdown. I created a service to shutdown the node based on the instructions at: https://askubuntu.com/questions/1449198/running-a-script-at-shutdown-before-networking-goes-down . I verified the shutdown service script does execute during computer shutdown using the time stamp as described in that article. I am concerned though, because when I shutdown the node it takes about half a minute, and the message appears not to shutdown the clean shutdown process before it is done, and when I shutdown the computer it does not wait for a clean shutdown to complete. Is there a node shutdown command that will block script execution until it completes? 

Is there a command that can be used to determine whether shutdown has completed that a script can use to spin in a polling loop until completion is indicated? 

Is there another way to block, and postpone computer shutdown until the node shutdown completes?

**Reply by andyoknen** [reply]
I understood your question - in theory, of course, you can ask in a loop if there is a running process, something like: `ps aux | grep pocketcoind`

I can also recommend using docker, which waits for the container to stop when it is turned off.
